### PR TITLE
Working with libraries in Echidna

### PR DESCRIPTION
**Pull request by ggrieco-tob**
This tutorial depends on the `--compile-libraries` which I believe will be in the next `crytic-compile` release

### CODE REVIEW

File: program-analysis/echidna/advanced/working-with-libraries.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+**Table of contents:**
+
+- [Introduction](#introduction)
+- [Example code](#example-code)
+- [Deploying libraries](#deploying-libraries)
+- [Linking libraries](#linking-libraries)
+
+## Introduction
+
+One important feature used Solidity introduced a brand new concept of smart contract library. While this feature is not extremely popular, a good amount of contracts use them to structure code and reduce the amount of bytecode deployed.
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply program-analysis/echidna/advanced/working-with-libraries.md:12]: Simplified.

File: program-analysis/echidna/advanced/working-with-libraries.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+Before continue, it is very recommended to review [the official Solidity documentation](https://docs.soliditylang.org/en/v0.8.19/contracts.html#libraries) to make sure
+you understand how libraries are created and deployed. When a user creates a library, Solidity will do one of these two options:
+
+- If all the functions are internal, the library is compiled into bytecode and added into the contracts that use it.
+- If there are some external functions, the library should be deployed into some address. Finally, the bytecode calling the library should be linked.
+
+If your libraries only contain internal functions, then Echidna will work correctly and you don't need to do anything extra to start your testing (you can skip the rest of the tutorial).
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply program-analysis/echidna/advanced/working-with-libraries.md:19]: Done

File: program-analysis/echidna/advanced/working-with-libraries.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+Before a contract can use a deployed library, its bytecode requires to be linked. This procedure requries to replace a particular string place holder
+in the bytecode by the address of the deployed library. Normally, either solc or the compilation framework (e.g. truffle) will take care of this.
+However, in our case, we will use `crytic-compile`, since it is easier to handle all cases from different frameworks just adding one new argument
+to pass to `crytic-compile` from Echidna:
+
+```yaml
+cryticArgs: ["--compile-libraries=(ConvertLib,0x1f)"]
+```
+
+Going back to the example, if we have both config options in a single config file (`echidna.yaml), we can run the metacoin contract
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply program-analysis/echidna/advanced/working-with-libraries.md:53]: Fixed

File: program-analysis/echidna/advanced/working-with-libraries.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+We can use the coverage report to verify that function using the library (`getBalanceInEth`) is not reverting:
+
+```
+ 28 | *   |     function getBalanceInEth(address addr) public view returns(uint){
+ 29 | *   |             return ConvertLib.convert(getBalance(addr),2);
+ 30 |     |     }
+```
+
+However, the code of library itself will not have their coverage displayed correctly:
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply program-analysis/echidna/advanced/working-with-libraries.md:68]: Fixed

File: program-analysis/echidna/advanced/working-with-libraries.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+```
+ 6 |     | library ConvertLib{
+ 7 |     |     function convert(uint amount, uint conversionRate) public pure returns (uint convertedAmount)
+ 8 |     |     {
+ 9 |     |             return amount * conversionRate;
+10 |     |     }
+11 |     | }
+```
+
+This is caused by the usage of `delegatecall` to execute contract code and unfortunately we do not have a workaround for it right now.
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply program-analysis/echidna/advanced/working-with-libraries.md:78]: A small summary section was added was added.

[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply program-analysis/echidna/advanced/working-with-libraries.md]: > Can we link to an open github issue for the user to track this.

Done.

> We should probably make a subsection "coverage limitation"

I will update the corresponding echidna tutorial with that.

**Comment by ggrieco-tob** [comment]
All the comments were addressed and all the tests are passing, except for this one:
```
FILE: ./not-so-smart-contracts/cosmos/non_determinism/README.md
[✖] https://lev.pm/posts/2020-04-18-golang-map-randomness/

9 links checked.

ERROR: 1 dead links found!
[✖] https://lev.pm/posts/2020-04-18-golang-map-randomness/ → Status: 404
```
but that's unrelated to this PR (it should be failing in master). @montyly please take another look or consider merge this one.

**Comment by montyly** [comment]
Note: We also need to update `SUMMARY.md`

**Pull request merged**